rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,777 @@
+            pkt_id: None,
+            complete: false,
+            logged: LoggerFlags::new(),
+            msg: Vec::new(),
+            toclient: false,
+            toserver: false,
+            de_state: None,
+            events: std::ptr::null_mut(),
+            detect_flags: applayer::TxDetectFlags::default(),
+        };
+        m.msg.push(msg);
+        return m;
+    }
+
+    pub fn free(&mut self) {
+        if self.events != std::ptr::null_mut() {
+            core::sc_app_layer_decoder_events_free_events(&mut self.events);
+        }
+        if let Some(state) = self.de_state {
+            core::sc_detect_engine_state_free(state);
+        }
+    }
+}
+
+impl Drop for MQTTTransaction {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+pub struct MQTTState {
+    tx_id: u64,
+    request_buffer: Vec<u8>,
+    response_buffer: Vec<u8>,
+    pub protocol_version: u8,
+    transactions: Vec<MQTTTransaction>,
+    connected: bool,
+}
+
+impl MQTTState {
+    pub fn new() -> Self {
+        Self {
+            tx_id: 0,
+            request_buffer: Vec::new(),
+            response_buffer: Vec::new(),
+            protocol_version: 0,
+            transactions: Vec::new(),
+            connected: false,
+        }
+    }
+
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if tx.tx_id == tx_id + 1 {
+                return Some(tx);
+            }
+        }
+        return None;
+    }
+
+    pub fn get_tx_by_pkt_id(&mut self, pkt_id: u32) -> Option<&mut MQTTTransaction> {
+        for tx in &mut self.transactions {
+            if let Some(mpktid) = tx.pkt_id {
+                if mpktid == pkt_id {
+                    return Some(tx);
+                }
+            }
+        }
+        return None;
+    }
+
+    fn new_tx(&mut self, msg: MQTTMessage, toclient: bool) -> MQTTTransaction {
+        let mut tx = MQTTTransaction::new(msg);
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        if toclient {
+            tx.toclient = true;
+        } else {
+            tx.toserver = true;
+        }
+        return tx;
+    }
+
+    fn handle_msg(&mut self, msg: MQTTMessage, toclient: bool) {
+        match msg.op {
+            MQTTOperation::CONNECT(ref conn) => {
+                self.protocol_version = conn.protocol_version;
+                if let Some(tx) = self.get_tx_by_pkt_id(MQTT_CONNECT_PKT_ID) {
+                    &mut MQTTState::set_event(tx, MQTTEvent::DoubleConnect);
```

Comment:
`self.get_tx_by_pkt_id` seems costly in time
Could we have a flag in the state ?
Or with the progress ?

Reply:
We actually already have a `connected` field in the state. Makes sense to use it here.